include/ibamr/WaveDampingStrategy.h

```diff
@@ -0,0 +1,119 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/////////////////////// INCLUDE GUARD ////////////////////////////////////
+
+#ifndef included_WaveDampingStrategy
+#define included_WaveDampingStrategy
+
+///////////////////////////// INCLUDES ///////////////////////////////////
+#include <string>
+
+#include "Variable.h"
+#include "ibamr/AdvDiffHierarchyIntegrator.h"
+#include "ibamr/INSVCStaggeredHierarchyIntegrator.h"
+#include "ibtk/ibtk_utilities.h"
+#include "tbox/Pointer.h"
+
+namespace IBAMR
+{
+/*!
+ * A collection of post processing call back functions to be hooked into
+ * IBAMR::HierarchyIntegrator class to employ a wave damping zone of some
+ * prescribed length within the computational domain. Additional strategies
+ * can be implemented as simple C functions within this collection.
+ *
+ * The instance of a WaveDampingStrategy and the particular damping function
+ * should be registered with an appropriate hierarchy integrator via
+ * registerPostprocessIntegrateHierarchyCallback.
+ *
+ * \param ctx is the pointer to WaveDampingStrategy struct.
+ */
+
+namespace WaveDampingFunctions
+{
+void callRelaxationZoneCallbackFunction(double current_time,
+                                        double new_time,
+                                        bool skip_synchronize_new_state_data,
+                                        int num_cycles,
+                                        void* ctx);
+
+void callConservedWaveAbsorbingCallbackFunction(double current_time,
+                                                double new_time,
+                                                bool skip_synchronize_new_state_data,
+                                                int num_cycles,
+                                                void* ctx);
+} // namespace WaveDampingFunctions
+
+/*!
+ * A struct holding the required information used by the wave damping function.
+ */
+struct WaveDampingStrategy
+{
+    /*
+     * Pointers to the fluid and advection-diffusion integrators.
+     */
+    SAMRAI::tbox::Pointer<IBAMR::INSVCStaggeredHierarchyIntegrator> d_ins_hier_integrator;
+    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> d_adv_diff_hier_integrator;
+
+    /*
+     * Pointer to the level set variable representing the wave interface.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_phi_var;
```

Comment:
Does this also need the context?

Reply:
I will also rename this to `WaveDampingData`

Reply:
`callConservedWaveAbsorbingCallbackFunction()` uses both current and new to solve a nonlinear mass-balance equation. It really is an implementation dependent thing :-(

Reply:
But the class that calls those functions presumably knows where it wants the results stored? This is what I want to make sure is codified in the API: where do the results go?

Reply:
(And similarly knows when it wants the results computed?)

Reply:
I see. In practice, I do it as postprocessing and it works great. I will just add new_context pointer.